File: Android/scenes/ktv/src/main/java/io/agora/scene/ktv/service/KTVServiceProtocol.kt
package io.agora.scene.ktv.service

import io.agora.rtc2.UserInfo
import io.agora.rtmsyncmanager.model.AUIRoomInfo
import io.agora.rtmsyncmanager.model.AUIUserInfo
import io.agora.rtmsyncmanager.model.AUIUserThumbnailInfo
import io.agora.scene.base.component.AgoraApplication
import io.agora.scene.ktv.KTVLogger
import io.agora.scene.ktv.widget.song.SongItem

/**
 * Ktv service listener protocol
 *
 * @constructor Create empty Ktv service listener protocol
 */
interface KtvServiceListenerProtocol {

    /**
     * On room expire
     *
     * @param channelName
     */
    fun onRoomExpire(channelName: String) {}

    /**
     * On room destroy
     *
     * @param channelName
     */
    fun onRoomDestroy(channelName: String) {}

    /**
     * On user list did changed
     *
     * @param userList
     */
    fun onUserListDidChanged(userList: List<AUIUserInfo>) {}

    /**
     * On mic seat snapshot
     *
     * @param seatMap
     */
    fun onMicSeatSnapshot(seatMap:Map<Int,RoomMicSeatInfo>){}

    /**
     * On current user audio mute
     *
     * @param userId
     * @param mute
     */
    fun onUserAudioMute(userId: String, mute: Boolean) {}

    /**
     * On current user video mute
     *
     * @param userId
     * @param mute
     */
    fun onUserVideoMute(userId: String, mute: Boolean) {}

    /**
     * On user enter seat
     *
     * @param seatIndex
     * @param userInfo
     */
    fun onUserEnterSeat(seatIndex: Int,userInfo: AUIUserThumbnailInfo) {}

    /**
     * On user leave seat
     *
     * @param seatIndex
     * @param userInfo
     */
    fun onUserLeaveSeat(seatIndex: Int,userInfo: AUIUserThumbnailInfo) {}

    /**
     * On update all choose songs
     *
     * @param chosenSongList
     */
    fun onChosenSongListDidChanged(chosenSongList: List<RoomSongInfo>) {}

    /**
     * On choristers list did changed
     *
     * @param choristerList
     */
    fun onChoristerListDidChanged(choristerList: List<RoomChoristerInfo>) {}

    /**
     * On chorister did enter
     *
     * @param chorister
     */
    fun onChoristerDidEnter(chorister: RoomChoristerInfo) {}

    /**
     * On chorister did leave
     *
     * @param chorister
     */
    fun onChoristerDidLeave(chorister: RoomChoristerInfo) {}
}

/**
 * Ktv service protocol
 *
 * @constructor Create empty Ktv service protocol
 */
interface KTVServiceProtocol {

    companion object {
        private val instance by lazy {
            // KTVServiceImp()
            KTVSyncManagerServiceImp(AgoraApplication.the()) { error ->
                error?.message?.let { KTVLogger.e("SyncManager", it) }
            }
        }

        fun getImplInstance(): KTVServiceProtocol = instance
    }

    /**
     * Reset
     *
     */
    fun reset()

    /**
     * Get room list
     *
     * @param completion
     * @receiver
     */
    fun getRoomList(completion: (list: List<AUIRoomInfo>) -> Unit)

    /**
     * Create room
     *
     * @param createRoomInfo
     * @param completion
     * @receiver
     */
    fun createRoom(createRoomInfo: CreateRoomInfo, completion: (error: Exception?, out: JoinRoomInfo?) -> Unit)

    /**
     * Join room
     *
     * @param inputModel
     * @param completion
     * @receiver
     */
    fun joinRoom(roomId: String, password: String?, completion: (error: Exception?, out: JoinRoomInfo?) -> Unit)

    /**
     * Leave room
     *
     * @param completion
     * @receiver
     */
    fun leaveRoom(completion: (error: Exception?) -> Unit)

    /**
     * Get all user list
     *
     * @param success
     * @param error
     * @receiver
     */
    fun getAllUserList(completion: (error: Exception?, list: List<AUIUserInfo>?) -> Unit)

    /**
     * Get user info
     *
     * @param userId
     * @return
     */
    fun getUserInfo(userId: String): AUIUserInfo?

    /**
     * Mute user audio
     *
     * @param mute
     * @param completion
     * @receiver
     */
    fun muteUserAudio(mute: Boolean, completion: (error: Exception?) -> Unit)

    /**
     * Mute user video
     *
     * @param mute
     * @param completion
     * @receiver
     */
    fun muteUserVideo(mute: Boolean, completion: (error: Exception?) -> Unit)

    /**
     * On seat
     *
     * @param seatIndex null autoOnseat
     * @param completion
     * @receiver
     */
    fun enterSeat(seatIndex: Int?, completion: (error: Exception?) -> Unit)

    /**
     * Out seat
     *
     * @param seatIndex
     * @param completion
     * @receiver
     */
    fun outSeat(seatIndex: Int, completion: (error: Exception?) -> Unit)

    /**
     * Get all song
     *
     * @param completion
     * @receiver
     */
    fun getAllSongList(completion: (error: Exception?, list: List<SongItem>) -> Unit)

    /**
     * Get chosen songs list
     *
     * @param completion`
     * @receiver
     */
    fun getChosenSongList(completion: (error: Exception?, list: List<RoomSongInfo>?) -> Unit)

    /**
     * Remove song
     *
     * @param songInfo
     * @param completion
     * @receiver
     */
    fun removeSong(songCode: String, completion: (error: Exception?) -> Unit)

    /**
     * Choose song
     *
     * @param songInfo
     * @param completion
     * @receiver
     */
    fun chooseSong(songInfo: RoomSongInfo, completion: (error: Exception?) -> Unit)

    /**
     * Make song top
     *
     * @param songCode
     * @param completion
     * @receiver
     */
    fun pinSong(songCode: String, completion: (error: Exception?) -> Unit)

    /**
     * Make song did play
     *
     * @param songCode
     * @param completion
     * @receiver
     */
    fun makeSongDidPlay(songCode: String, completion: (error: Exception?) -> Unit)

    /**
     * Join chorus
     *
     * @param songCode
     * @param completion
     * @receiver
     */
    fun joinChorus(songCode: String, completion: (error: Exception?) -> Unit)

    /**
     * Leave chorus
     *
     * @param completion
     * @receiver
     */
    fun leaveChorus(completion: (error: Exception?) -> Unit)

    /**
     * Get chorister list
     *
     * @param completion
     * @receiver
     */
    fun getChoristerList(completion: (error: Exception?, choristerInfoList: List<RoomChoristerInfo>?) -> Unit)

    /**
     * Subscribe listener
     *
     * @param listener
     */
    fun subscribeListener(listener: KtvServiceListenerProtocol)

    /**
     * Unsubscribe listener
     *
     * @param listener
     */
    fun unsubscribeListener(listener: KtvServiceListenerProtocol)

    /**
     * Get seat map
     *
     * @return
     */
    fun getSeatMap():Map<Int,RoomMicSeatInfo>{
        return emptyMap()
    }

    /**
     * Get chorister list
     *
     * @return
     */
    fun getChoristerList():List<RoomChoristerInfo>{
        return emptyList()
    }

    /**
     * Get chosen song list
     *
     * @return
     */
    fun getChosenSongList(): List<RoomSongInfo>{
        return emptyList()
    }
}

File: iOS/AgoraEntScenarios/Scenes/KTV/Service/KTVServiceProtocol.swift
//
//  KTVServiceProtocol.swift
//  AgoraEntScenarios
//
//  Created by CP on 2023/7/13.
//

import Foundation
import RTMSyncManager

@objc protocol KTVServiceListenerProtocol: NSObjectProtocol {
    
    /// The room is expired.
    func onRoomDidExpire()
    
    /// The room was destroyed.
    func onRoomDidDestroy()
    
    /// Change in the number of room users
    /// - Parameter userCount: Number of users
    func onUserCountUpdate(userCount: UInt)
    
    /// Update of the seat
    /// - Parameter seat: Seat object
    func onMicSeatSnapshot(seat: [String: VLRoomSeatModel])
    
    /// Update of the seat
    /// - Parameter seat: Seat object
    func onUserSeatUpdate(seat: VLRoomSeatModel)
    
    /// There are members on the Seat (take the initiative to go to the seat/the landlord hugs people to the seat)
    /// - Parameters:
    ///   - seatIndex: Seat index
    ///   - user: User information
    func onUserEnterSeat(seatIndex: Int, user: AUIUserThumbnailInfo)
    
    /// Some members go down the seat (take the initiative to go down the seat/homeowner kicks people down the seat)
    /// - Parameters:
    ///   - seatIndex: Seat index
    ///   - user: User information
    func onUserLeaveSeat(seatIndex: Int, user: AUIUserThumbnailInfo)
    
    /// The owner muted/lifted the ban on the seat.
    /// - Parameters:
    ///   - seatIndex: Seat index
    ///   - isMute:Microphone switch status
    func onSeatAudioMute(seatIndex: Int, isMute: Bool)

    /// The owner disables/enables/enables the seat camera
    /// - Parameters:
    ///   - seatIndex: Seat index
    ///   - isMute: Camera switch status
    func onSeatVideoMute(seatIndex: Int, isMute: Bool)
    
    /// Update all song callbacks (such as pin)
    /// - Parameter song: List of songs
    func onChosenSongListDidChanged(songs: [VLRoomSelSongModel])
    
    /// The chorus joins
    /// - Parameter chorus: Joined chorus information
    func onChoristerDidEnter(chorister: KTVChoristerModel)
    
    /// The chorus left.
    /// - Parameter chorister: The chorus who left
    func onChoristerDidLeave(chorister: KTVChoristerModel)
    
}

@objc protocol KTVServiceProtocol: NSObjectProtocol {
    // room info
    
    /// Get the list of rooms
    /// - Parameters:
    ///   - page: Page number
    ///   - completion: Complete the callback
    func getRoomList(page: UInt, completion: @escaping (Error?, [AUIRoomInfo]?) -> Void)
    
    /// Create a room
    /// - Parameters:
    /// - inputModel: Input Model
    /// - completion: Complete the callback
    func createRoom(inputModel: KTVCreateRoomInfo, completion: @escaping (Error?, AUIRoomInfo?) -> Void)
    
    /// Join the room
    /// - Parameters:
    /// - roomId: room id
    /// - password: password
    /// - completion: Complete the callback
    func joinRoom(roomId: String, password: String, completion: @escaping (Error?) -> Void)
    
    /// Leave the room
    /// - Parameter completion: Complete the callback
    func leaveRoom(completion: @escaping (Error?) -> Void)
    
    // mic seat
    
    /// Enter Seat
    /// - Parameters:
    ///   - seatIndex: Seat Index
    ///   - completion: Complete the callback
    func enterSeat(seatIndex: NSNumber?, completion: @escaping (Error?) -> Void)
    
    /// Leave Seat
    /// - Parameters:
    ///   - completion: Complete the callback
    func leaveSeat(completion: @escaping (Error?) -> Void)
    
    /// Kick seat
    /// - Parameters:
    ///   - seatIndex: Seat index
    ///   - completion: Complete the callback
    func kickSeat(seatIndex: Int, completion: @escaping (NSError?) -> ())
    
    /// Set seat voice
    /// - Parameters:
    ///   - muted: Whether to turn off the sound, YES: turn off the sound, NO: turn on the sound
    ///   - completion: Complete the callback
    func updateSeatAudioMuteStatus(muted: Bool, completion: @escaping (Error?) -> Void)
    
    /// Set seat camera state
    /// - Parameters:
    ///   - muted: Whether to turn off the camera, YES: turn off the camera, NO: turn on the camera
    ///   - completion: Complete the callback
    func updateSeatVideoMuteStatus(muted: Bool, completion: @escaping (Error?) -> Void)
    
    // choose songs
    
    /// Delete the selected song
    /// - Parameters:
    /// - songCode: Song id
    /// - completion: Complete the callback
    func removeSong(songCode: String, completion: @escaping (Error?) -> Void)
    
    /// Get the list of selected songs
    /// - Parameter completion: Complete the callback
    func getChoosedSongsList(completion: @escaping (Error?, [VLRoomSelSongModel]?) -> Void)
    
    /// The lead singer tells the songs currently playing in the background
    /// - Parameters:
    /// - songCode: room id
    /// - completion: Complete the callback
    func markSongDidPlay(songCode: String, completion: @escaping (Error?) -> Void)
    
    /// order a song
    /// - Parameters:
    /// - inputModel: Input Model
    /// - completion: Complete the callback
    func chooseSong(inputModel: KTVChooseSongInputModel, completion: @escaping (Error?) -> Void)
    
    /// Pin the top song
    /// - Parameters:
    /// - songCode: Song id
    /// - completion: Complete the callback
    func pinSong(songCode: String, completion: @escaping (Error?) -> Void)
    
    /// Join the chorus
    /// - Parameters:
    /// - songCode: Song id
    /// - completion: Complete the callback
    func joinChorus(songCode: String, completion: @escaping (Error?) -> Void)
    
    /// Cancel the chorus of the accompaniment
    /// - Parameters:
    /// - songCode: Song id
    /// - completion: Complete the callback
    func leaveChorus(songCode: String, completion: @escaping (Error?) -> Void)
    
    /// Subscribe to the listener
    /// - Parameter listener: Monitor object
    func subscribe(listener: KTVServiceListenerProtocol?)
    
    /// Get the current room usage time
    /// - Parameter channelName: Channel Name
    /// - Returns: Duration of use, unit ms
    func getCurrentDuration(channelName: String) -> UInt64
}
